### PR TITLE
term: next gen, phase one

### LINKED ISSUE #353: prompt doesn't always update
upon the pier first starting, after ~&s and idle ^C, etc.

### LINKED ISSUE #2646: unused Dill tasks
There are four `task`s in Dill that appear to be unused.

`$flow` and `$harm` appear in `+task:able:dill` in `zuse.hoon`, but their entry in `+call` in `dill.hoon` is just a lark expression that does nothing.

`$hook` and `$talk` also appear in `+task:able:dill` in `zuse.hoon` but have no entry in `+call` in `dill.hoon`.

Talking to @philipcmonk, these might be a mixture of `task`s that once did something but have been deprecated, or unfinished `task`s that someone intends to get to eventually, so I don't think they necessarily all need to be scrubbed.

### PR DESCRIPTION
**Pull request by Fang-**
Bunch of small goodies in here, but the majority of this revolves around decoupling term.c, dill and drum.

Previously, they all knew about and played a role in drum's "prompt at the bottom" semantics.  
With these changes, only drum knows about the prompt. term.c just remembers whatever the line at the bottom of the screen contains to aid in spinner drawing. It also whimsically inserts printfs between the second-to-last and bottom lines of the screen, but those only show up in the default session, which is always linked to drum, so it works out fine.

Other fixes and improvements include:
- Support for 2D cursor movement, and by extension 2D screen printing,
- Better (but still incomplete) support for multiple sessions in dill,
- Minimal X10-style mouse-click support,
- A cleaned up dill api,
- General cleanup and smaller improvements.

There's some remaining work here, but this is probably good to go for a first review while I work on that. While the _order_ of these commits is a bit all over the place, the commits _themselves_ should be well-contained changes, often paired with additional commentary, and as such probably the best way to review all of this.  
Please also keep an eye out for `//REVIEW` comments strewn about, I'd love some input on those.

Some remaining work:
- [x] Fix/improve `-t` mode behavior. See also CI output, it's kinda not good in some places.
- [x] ~~Update webterm and its dependencies to reflect these changes. (The blit mark doesn't even build rn, _but CI passes anyway!_ [distant screaming])~~ Updated, but not yet perfect. Remaining work includes:
  - [x] Restructuring and cleanup of codebase, follow best practices.
  - [x] Improved subscription management. (Pending broader landscape subscription reform.)
  - [x] Respond to window resizing.
  - [x] Respond to landscape theme (light vs dark).
- [x] Update KH to reflect these changes.
- [x] Make clay and gall send proper printout requests to dill, instead of having dill handle their updates locally.
- [x] ~~Experiment with multi-session dill in webterm, settle on an interface. May or may not get~~ Settled on an (as of yet unreviewed) interface. Further drum & webterm work punted to a separate PR. Fixes #2646.

Currently targeting ~`release/next-sys`~ `release/next-dill`, which is the merge of `next-sys` and `next-js`, so we have a cleaner diff here.  
The arvo-side changes depend on the runtime-side changes. So maybe we want `release/next-step` or w/e for that? Also depends on urbit/landscape#188 if we want users to actually know about the strict need for a runtime upgrade.

Might fix #353, if that wasn't already the case.

### CODE REVIEW

**Comment by Fang-** [comment]
Noting for my own records that there's currently a bug in this, where sometimes when printing large amounts of output (ie, printing many lines), the last handful of lines have double newlines between them, where there should be none. Suspecting that this relates to the spinner logic, because I hadn't seen this prior to those changes, but that's really just a hunch...

**Review by joemfb — CHANGES_REQUESTED** [review]
Sorry for the delay! I've noted a couple minor issues:

**Comment by Fang-** [comment]
Alright, nu-webterm is in. xterm.js is pretty nice to use, we get to basically just port the simple parts of `term.c` and call it a day. Still needs some work though. It doesn't handle window resizing at all nicely, and I feel like there's general structural improvements to be made.

@liam-fitzgerald please take a look. I might need your help with cleaning up/restructuring the single-file mess I've made, maybe we can pair a little bit if you have time.

File: pkg/arvo/lib/hood/drum.hoon
Comment (Fang-) [reply pkg/arvo/lib/hood/drum.hoon:553]: Good, and done, but noting that this turns out kinda ugly due to #2688.

**Comment by Fang-** [comment]
Haskell changes are in, paging @pilfer-pandex for review. I haven't put the re/connecting to a running ship to the test yet, so don't be surprised if anything about that is wrong.

But also, if I understand correctly, it's storing history and such so that it can redraw the screen when you connect. But there's a dill move for that, which will go into the terminal handler and just redraw the screen for you, no further runtime logic required.  
(This doesn't account for slogs and logging output and such, and for current drum only redraws the prompt, but that will improve as more output gets reified and terminal handler technology evolves.)

**Comment by ohAitch** [comment]
Any roadmap on sixel support? :)

**Comment by Fang-** [comment]
@ohAitch I've been keeping an eye on xtermjs/xterm.js#2503 for laughs, since webterm will likely be the easiest place to experiment with these things, but it's not on the roadmap per se. It's certainly a nice-to-have, but there's plenty of other work in this area that's more profitable.

fwiw, I'm thinking the order of operations here is going to be something like:
1. this pr,
2. full session support in drum, webterm, eventually the runtimes,
3. drum-alternative with per-application buffers,
4. native scroll/back support & copy handling in drum-style buffers, enabling full mouse integration,
5. terminal-based experiments in the direction of "ui protocol"

5 is pretty underdefined, but sixels would happen somewhere around there. All this is of course somewhat tentative, and notably lower-priority work than many other things going on, so the best answer I can give you is "likely, eventually".

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs:88]:
```suggestion
    EvEdit t | (0, _) <- sCurPos -> st { sLine = t }
             | otherwise         -> st
```
and likewise below. This is called a "pattern guard." In general the existence of "guards" means you shouldn't ever need to put an if expression to the right hand side of a pattern match

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs:116]: isn't this just `map EvEdit sLine`?

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs:162]: Please line these up (or don't)

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Render.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Term/Render.hs:37]: no need for `$` here

File: pkg/hs/urbit-noun-core/lib/Urbit/Noun/TH.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-noun-core/lib/Urbit/Noun/TH.hs:101]: Awesome!

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:314]: man dat `($(`

it's like `(~(` in hoon

I'll see myself out

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:328]:
```suggestion
    Bol b -> toNoun b
```

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:317]: Suggested code:

- parse the noun coming in into a Maybe Bolt
- if it's just (use `maybe`, possibly), wrap in Bol
- otherwise, call the derived Belt parser

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs:117]: Some people would say that you should use a guard here instead of an if expression, but I'll leave it up to you

**Review by pilfer-pandex — COMMENTED** [review]
Looking mostly good on the haskell side, with a few suggested changes. Unfortunately, you now know more than me about the king's terminal driver, so my ability to, e.g., comment on the ANSI stuff is limitied

might be a good idea for you to teach me when I'm a bit less busy, if you want :)

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:317]: Actually, you ought to be able to accomplish the above using the Alternative instance of parsers. It should look something like

```
parseNoun n = Bol <$> parseNoun n <|> $(deriveFromNounFunc ''Belt) n
```

which also very clearly and cleanly expresses what is going on

File: pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs
Comment (Fang-) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Term/Logic.hs:116]: No? `sLine` isn't a list, so it doesn't let me `map` over it. Also, we want to explicitly produce `[]` for the null/empty string case, not `[EvEdit null]` or w/e.

File: pkg/interface/src/logic/api/term.ts
Comment (Fang-) [reply pkg/interface/src/logic/api/term.ts:6]: Just an `if` check or similar on the type/properties of the value.

(Of course in practice we only care about _constructing_ this input type, we don't get it sent to us.)

File: pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs
Comment (Fang-) [reply pkg/hs/urbit-king/lib/Urbit/Arvo/Event.hs:317]: > which also very clearly and cleanly expresses what is going on

image

**Comment by Fang-** [comment]
Alright, afaic this is good to go. Paging @pilfer-pandex, @liam-fitzgerald and @joemfb for final review on the Haskell side, webterm, and vere/dill/drum respectively.

Sessions _support_ is partially here, and will be completed in a subsequent PR. Webterm subscription management still does the old thing because the new thing isn't yet in. Any additional changes that need to be made to accommodate the release of this work will go in a separate PR also.

I've been running this on my local dev ships for the entire time I've worked on this. As far as I've been able to discern, no serious bugs remain.

**Comment by ryjm** [comment]
I've been running this on a fake zod for the past couple days, also not seeing anything strange. Anything I should be on the lookout for?

**Comment by Fang-** [comment]
@ryjm neat, thank you! Things I'm most concerned about are the refactored pre- and post-spinner logic, and interactions between arvo-level and runtime-level output. ie printfs overwriting parts of the screen, stuff like that.

**Comment by Fang-** [comment]
I don't think any of the typescript errors here are about webterm code, fwiw.

**Comment by matildepark** [comment]
> I don't think any of the typescript errors here are about webterm code, fwiw.

They aren't, there's about 127 TS errors on master and the work to clear the test (https://github.com/urbit/urbit/pull/4899) hasn't hit master. You could always pull in next-userspace to resolve that, unless this is going out sometime sooner than the next week or so.

File: pkg/arvo/sys/lull.hoon
Comment (philipcmonk) [inline pkg/arvo/sys/lull.hoon:1069]: Are we supposed to not be able to represent eg "control alt C"?  Feels like this should be `key=belt`

Comment (philipcmonk) [reply pkg/arvo/sys/lull.hoon:1069]: Or maybe just `mod=(set ?(%ctl %met %hyp))`

Comment (Fang-) [reply pkg/arvo/sys/lull.hoon:1069]: `%hyp` here represents the simultaneous pressing of `%ctl` and `%met`, not a single "hyper key".

Comment (philipcmonk) [reply pkg/arvo/sys/lull.hoon:1069]: Ah, okay

**Review by pilfer-pandex — CHANGES_REQUESTED** [review]

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:142]: This can (and should) be defined outside the scope of this function

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:168]: again, doesn't need to be defined in-component

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:261]: `Belt[]`

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:342]: consider refactoring into `inputToBelts` (where the meat of the logic is and can be outside component scope) and then `onInput` would just call that and forward the results to the api.

**Review by liam-fitzgerald — CHANGES_REQUESTED** [review]
some style and performance nits and will need a slight refactor to fit into the new api model. Else looks good!

File: pkg/interface/src/types/term-update.ts
Comment (Fang-) [reply pkg/interface/src/types/term-update.ts:12]: I didn't know about the distinction, so I read about these a little bit. It seems `interface` is _slightly_ fancier than `type`? The former seems to carry more OOP semantics with it than the latter, certainly.

I'm fine going either way, but I'm just defining some data's shape here, so it's not obvious to my why I'd want `interface` for that.

**Review by pilfer-pandex — APPROVED** [review]

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:282]: No need to use the `useTermState` hook you already have all these variables

File: pkg/interface/src/views/apps/term/app.tsx
Comment (liam-fitzgerald) [inline pkg/interface/src/views/apps/term/app.tsx:288]: We now have `useDark()` in `state/join.tsx` that does this logic for you

**Review by liam-fitzgerald — APPROVED** [review]
LGTM! Minor nits, nothing severe

File: pkg/urbit/include/vere/vere.h
Comment (joemfb) [inline pkg/urbit/include/vere/vere.h:148]: I believe we've discussed this before, but I've forgotten: why is the bottom line stored in the driver state as a noun instead of a processed `uv_buf_t` or whatever?

**Review by joemfb — COMMENTED** [review]
Some requests for clarification and minor nits:

File: pkg/urbit/include/vere/vere.h
Comment (Fang-) [reply pkg/urbit/include/vere/vere.h:148]: Because a blit might _partially_ overwrite the line. Using a `uv_buf_t` that would of course still be trivial for the plaintext case, but keeping _styling_ correct when that's in play would require parsing the `uv_buf_t` for escape sequences etc etc.

Comment (joemfb) [reply pkg/urbit/include/vere/vere.h:148]: I see, I hadn't realized that `mir.lin` isn't just the last blit, but has been processed with `+wail:klr:format` and `+pact:klr:format`.

File: pkg/arvo/sys/vane/gall.hoon
Comment (Fang-) [reply pkg/arvo/sys/vane/gall.hoon:284]: > can you write a little here about the reason for this %gall change?

I interpreted `%onto` as something gall was using solely to message dill about app load/load failures it wanted to be printed, and figured that should be replaced with gall asking dill explicitly to print some text.

Some more commentary in 5ef68d3.

> looks like drum changes state on `%onto`

Oh, I thought it was just partaking in upgrade notification printing, but on closer inspection it _appears_ to be doing some internal state song-and-dance too... Guessing this is the mechanism that auto-`|link`s started apps? But will have to play with this a little bit to see what exactly is affected.

> @belisarius222 was planning some overlapping changes

We already dm'ed about this a little bit, and the sentence "gall could deliver this info to dill with an %onto, or whatever" was used. (^: So, we'll need to talk about this a bit more I guess. I'm amenable to putting %onto back in if it turns out gall wants update status notifications for non-printing purposes.

File: pkg/urbit/include/vere/vere.h
Comment (joemfb) [reply pkg/urbit/include/vere/vere.h:148]: On second thought, I'm slightly less convinced. `+wail` as called is basically for line-wrapping, but it assumes that each `@c` is an individual glyph. This assumption is usually correct, but not always (see "grapheme clusters" ...). I'm not sure how much of a problem this is, but I wonder if we should bite the bullet and pin the spinner to the bottom right corner, or even a temporary new line.

Comment (Fang-) [reply pkg/urbit/include/vere/vere.h:148]: > or even a temporary new line

(Not sure that would work in the first place, but) honestly if we're going there we should just add a persistent status bar... and write a handful of `//TODO`s for things to put in it.

File: pkg/urbit/vere/io/term.c
Comment (Fang-) [reply pkg/urbit/vere/io/term.c:1435]: This is actually a little bit complicated. Printing it would give it the same semantics that `%put` has, requiring the sender to be aware of where the cursor is, make sure the url isn't overwriting anything important on-screen, and/or print a newline to make space for it.

Maybe that's fine, and we want to treat it that way because we can't guarantee the url will be opened. Thinking about the ideal future where all clients support all the blits fully, that's not what we want `%url` to do though.

The alternative is pushing responsibility for visually printing the url to whoever is emitting the `%url`.

Comment (Fang-) [reply pkg/urbit/vere/io/term.c:1435]: Looking into this again, perhaps we can get away with a simple `system()` depending on which OS we're building for. `open $url` on MacOS, `xdg-open $url` on Linux, `start $url` (or similar) on Windows.

But maybe there's some security concern there around apps being able to both write to disk (`%sav`) and open arbitrary strings.

Comment (Fang-) [reply pkg/urbit/vere/io/term.c:1435]: @philipcmonk the thing I mentioned during our last call, we actually already do:

https://github.com/urbit/urbit/blob/d138ab254b9985a1fdf50cdbdcf78e6f7fd1479e/pkg/arvo/lib/hood/drum.hoon#L916

That is, I was going to update drum to always print the url, in addition to sending a `%url` blit. Turns out I already did this five(!) months ago in 0d8288e.

File: pkg/arvo/sys/vane/gall.hoon
Comment (Fang-) [reply pkg/arvo/sys/vane/gall.hoon:284]: Looked into drum's `%onto` handling, and I can't quite make out what the `%&` case was doing. Auto-linking after `|start`, `|link` and `|unlink`ing, and `|fade`ing all work fine, and hit the same code paths they did before (with the exception of `+take-onto` itself, of course).

With that, I'd be reasonably comfortable calling the removal of `%onto` handling in drum a cosmetic change.

File: pkg/urbit/include/vere/vere.h
Comment (joemfb) [reply pkg/urbit/include/vere/vere.h:148]: Having thought about this some more, I'm comfortable with the status quo. The spinner as designed simply conflicts with a 2-dimensional output array, and that will have to be solved at some point. Since it's purely a feature of the runtime, it can be changed without impact to the %dill protocol.

**Review by joemfb — CHANGES_REQUESTED** [review]
@Fang- can you merge master into `release/next-dill`, and then update this PR to resolve the conflicts?

**Comment by joemfb** [comment]
This PR is now ready, the only thing still blocking it is an issue with the mingw windows build.

We're hitting an assertion in libuv, but only when running with `-t` (or -d`):

```
Assertion failed: req->write_buffer.base, file src/win/pipe.c, line 1063
```

Notably, this is the same assertion that we would hit if we were building against an unpatched libuv. @locpyl-tidnyd, do you have any insight into what might be going wrong here?

File: pkg/urbit/vere/io/term.c
Comment (Fang-) [inline pkg/urbit/vere/io/term.c:669]: @joemfb I don't think these are necessary? Current dill, as it exists on the network, is already forward-compatible with the inputs from the new runtime. See also #4942.

Comment (joemfb) [reply pkg/urbit/vere/io/term.c:669]: But previous dill isn't, so people with old piers could end up in a situation where they have to juggle binaries to get upgraded (or even just poke around). (Since non-modified `%txt` input will still work, they can probably get unstuck by just typing `|merge ...` -- these concerns are a little more theoretical.)

In general, dropping support for old protocols should happen after some longer period time, and it should be straightforward to continue supporting multiple versions. But those patterns are yet to be established, which is why I'm settling for comments.

Comment (Fang-) [reply pkg/urbit/vere/io/term.c:669]: Fair game. I'm relatively comfortable (and relieved (^: ) to merge this as is, but if any poor sap happens to get hurt by this: blame me.

**Comment by tacryt-socryp** [comment]
So how do you use any of the new terminal features such as session support?

**Comment by Fang-** [comment]
@tacryt-socryp Depends on what context you're talking about.

Drum/sole/shoe agents can distinguish between connecting sessions from the subscription path. For `+on-watch` calls where the path starts with `/sole`, you can call `+path-to-id` from the `/lib/sole` to receive a `sole-id` identifying the session. Use this to register it in your state. Inputs for that session will have that same id in their `%sole-action` pokes. If you use `/lib/shoe`, these ids just come in as arguments on the shoe-specific arms.  
The rest of the sole protocol has not changed. You do not get any new output affordances.

If you're developing a standalone application, it's possible for you to speak the dill protocol directly, like drum does. If a session connects to an agent directly, it gets watched on a `/dill/[session-id]` path. For inputs, it will receive pokes with a `%dill-poke` mark, containing a `[session=@ta belt=dill-belt:dill]`. To send outputs, produce `%dill-blit` facts containing a single `blit:dill`.  
The `blit:dill` includes new affordances, like moving the cursor around on the screen, to draw at arbitrary locations. It's also very primitive, in practice this means you'll write a lot of dumb rendering logic.

The web terminal can create and connect to different sessions. To connect directly to a custom agent, use a session name with the following format: `agent!session-name`. If the target agent does not support the dill protocol, the session will be unusable.  
I have ideas on how to let apps bring you into specific webterm sessions like this in a more user-friendly way, stay tuned.

The runtime currently does not support connecting to anything but the default session, as it always has.

**Pull request merged**
